File: apps/admin-v2/src/infrastructure/api/repositories/api-service.repository.ts
import type {
  ServiceRepository,
  CreateServiceData,
  ListServicesParams,
} from '@/domain/repositories/service.repository';
import type { Service } from '@/domain/entities/service';
import type { PaginatedResult } from '@/shared/types/api';
import api from '../client';
import { mapService } from '../mappers/service.mapper';
import { mapPaginatedResponse } from '../mappers/pagination';

export class ApiServiceRepository implements ServiceRepository {
  async getAll(params?: ListServicesParams): Promise<PaginatedResult<Service>> {
    // Sin parámetros el que llama quiere el catálogo entero. Los selectores
    // dependen de eso: paginarlos escondería servicios sin que nadie se entere,
    // que es exactamente el bug que el paginado no debe reintroducir.
    const paged = Boolean(params?.page || params?.perPage || params?.q);

    const query: Record<string, unknown> = paged
      ? {
          page: params?.page ?? 1,
          ...(params?.perPage ? { per_page: params.perPage } : {}),
          ...(params?.q ? { q: params.q } : {}),
        }
      : { per_page: 'all' };

    const { data: res } = await api.get('/services', { params: query });
    return mapPaginatedResponse(res, mapService);
  }

  async getById(id: string): Promise<Service> {
    const { data: res } = await api.get(`/services/${id}`);
    return mapService(res.data);
  }

  async create(data: CreateServiceData): Promise<Service> {
    const { data: res } = await api.post('/services', {
      name: data.name,
      price: data.price,
      description: data.description,
      image_url: data.imageUrl || null,
      is_active: data.isActive,
      staffing: data.staffing ?? 'washer',
      sort_order: data.sortOrder,
    });
    return mapService(res.data);
  }

  async update(id: string, data: Partial<CreateServiceData>): Promise<Service> {
    const body: Record<string, unknown> = {};
    if (data.name !== undefined) body.name = data.name;
    if (data.price !== undefined) body.price = data.price;
    if (data.description !== undefined) body.description = data.description;
    if (data.imageUrl !== undefined) body.image_url = data.imageUrl || null;
    if (data.isActive !== undefined) body.is_active = data.isActive;
    if (data.staffing !== undefined) body.staffing = data.staffing;
    if (data.sortOrder !== undefined) body.sort_order = data.sortOrder;

    const { data: res } = await api.put(`/services/${id}`, body);
    return mapService(res.data);
  }

  async delete(id: string): Promise<void> {
    await api.delete(`/services/${id}`);
  }
}

File: apps/admin_mobile/lib/infrastructure/api/repositories/api_service_repository.dart
import '../../../domain/entities/service.dart';
import '../../../domain/repositories/service_repository.dart';
import '../../../shared/types/paginated_result.dart';
import '../dio_client.dart';
import '../mappers/service_mapper.dart';
import 'paginated_helper.dart';

class ApiServiceRepository implements ServiceRepository {
  final DioClient _client;

  ApiServiceRepository(this._client);

  @override
  Future<PaginatedResult<Service>> getAll({int? page}) async {
    final params = <String, dynamic>{};
    if (page != null) params['page'] = page;

    final response =
        await _client.dio.get('/services', queryParameters: params);
    return extractPaginated(response.data, mapService);
  }

  @override
  Future<Service> create({
    required String name,
    required double price,
    String? description,
    String? imageUrl,
    bool? isActive,
  }) async {
    final body = <String, dynamic>{
      'name': name,
      'price': price,
    };
    if (description != null) body['description'] = description;
    if (imageUrl != null) body['image_url'] = imageUrl;
    if (isActive != null) body['is_active'] = isActive;

    final response = await _client.dio.post('/services', data: body);
    final data = response.data['data'] ?? response.data;
    return mapService(data as Map<String, dynamic>);
  }

  @override
  Future<Service> update(
    int id, {
    String? name,
    double? price,
    String? description,
    String? imageUrl,
    bool? isActive,
    int? sortOrder,
  }) async {
    final body = <String, dynamic>{};
    if (name != null) body['name'] = name;
    if (price != null) body['price'] = price;
    if (description != null) body['description'] = description;
    if (imageUrl != null) body['image_url'] = imageUrl;
    if (isActive != null) body['is_active'] = isActive;
    if (sortOrder != null) body['sort_order'] = sortOrder;

    final response = await _client.dio.put('/services/$id', data: body);
    final data = response.data['data'] ?? response.data;
    return mapService(data as Map<String, dynamic>);
  }

  @override
  Future<void> delete(int id) async {
    await _client.dio.delete('/services/$id');
  }
}
